### PR TITLE
Feature/pql validation

### LINKED ISSUE #12: Introduce PQL validation
Introduce a layer of schema validation to ensure that the PQL request is well formed before parsing it to the PQL parser.  This will allow the app to fail fast if the PQL request is badly formed, reduce the risk of errors from edge cases that haven't been accounted for in downstream code and also allow for centralisation of correctness logic.  e.g. instead of having to perform these checks in multiple places: https://github.com/paralink-network/paralink-node/blob/master/src/pql/parser.py#L63, https://github.com/paralink-network/paralink-node/blob/master/src/pql/pipeline.py#L70, https://github.com/paralink-network/paralink-node/blob/master/src/pql/pipeline.py#L90, https://github.com/frisitano/paralink-node/blob/feature/arbitrary_user_query/src/pql/query_sql.py#L27 etc.

### PR DESCRIPTION
**Pull request by frisitano**
Introduce PQL schema validation using the [`jsonschema`](https://python-jsonschema.readthedocs.io/en/stable/) library which  implements the [json-schema](https://json-schema.org/) standard.  The implementation decomposes PQL into subschemas and then combines them in a modular fashion to create a comprehensive schema for PQL.  

Closes #12

### CODE REVIEW

File: src/pql/schema.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    "type": "object",
+    "properties": {"step": {"enum": []}},
+    "required": ["step"],
+}
+
+step_query_sql = {
+    "type": "object",
+    "properties": {
+        "step": {"const": "query.sql"},
+        "method": {"enum": ["json", "list", None]},
```
Comment (jbargu) [inline src/pql/schema.py:98]: What does it mean for method to be `None`?

File: src/pql/schema.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    "properties": {
+        "step": {"const": "math"},
+        "method": {"enum": ["mul", "add", "sub", "div"]},
+        "params": {"type": "number"},
+        "direction": {"const": {"reverse"}},
+    },
+    "required": ["step", "method", "params"],
+}
+
+step_custom = {
```
Comment (jbargu) [inline src/pql/schema.py:88]: I believe this has to be documented somehow, probably best if the custom step schema has to be supplied along with the definition in `src.config.PQL_CUSTOM_METHODS` and then the `source` object below also appends the schemas found in this object. Maybe even better if a custom function is defined as a `CustomFunction` object or something similar to make it more readable. What do you think?

Right now the `examples/custom_function.py` won't work due to missing schema.

**Review by jbargu — CHANGES_REQUESTED** [review]
This looks quite good! It is very clean, I just have a few remarks. Like we discussed, I would remove the syntax checking across the code that is deprecated by this PR. More insightful message would also be very much appreciated, however I understand that this can be a problem.

File: src/pql/schema.py (hunk 1 above)
Comment (frisitano) [reply src/pql/schema.py:95]: This corresponds to the case in which we do not need to parse the data before we push it into the SQLite table as defined by [this](https://github.com/frisitano/paralink-node/blob/feature/pql_validation/src/pql/query_sql.py#L24) condition.

File: src/pql/schema.py (hunk 2 above)
Comment (frisitano) [reply src/pql/schema.py]: I've made an attempt at your proposed implementation.  Good idea 👍

File: src/pql/exceptions.py [hunk 3]
```diff
@@ -87,3 +86,12 @@ def __init__(self, message: str):
         self.code = -32015
         self.message = message
         self.data = UNSET
+
+
+class PqlValidationError(Error):
+    """PqlValidationError is triggered when the Pql fails validation"""
+
+    def __init__(self, message: str):
+        self.code = -32015
```
[the remark this answers is not in the document]
Comment (frisitano) [reply src/pql/exceptions.py]: Good catch.

File: src/pql/schema.py (hunk 2 above)
Comment (frisitano) [reply src/pql/schema.py]: I've created an alternative implementation that allows the user to specify the custom methods / plugs as text and then we import using importlib.  You can find that implementation [here](https://github.com/frisitano/paralink-node/blob/sklearn/src/config.py#L63).

Comment (jbargu) [reply src/pql/schema.py]: I like the multi-file approach with it. It is up to you, if you fancy it more, just submit another PR.

File: src/pql/pipeline.py [hunk 4]
```diff
@@ -115,27 +99,23 @@ async def math(self, step: dict, index: int) -> Decimal:
...
+        calculations: typing.Dict[
+            tuple, typing.Callable[[Decimal, Decimal], Decimal]
+        ] = {
+            ("mul", "forward"): lambda curr, params: curr * params,
+            ("add", "forward"): lambda curr, params: curr + params,
+            ("sub", "forward"): lambda curr, params: curr - params,
+            ("sub", "reverse"): lambda curr, params: params - curr,
+            ("div", "forward"): lambda curr, params: curr / params,
+            ("div", "reverse"): lambda curr, params: params / curr,
+        }
```
Comment (jbargu) [inline src/pql/pipeline.py:111]: Very elegant and clean. :dancers:

**Comment by jbargu** [comment]
This PR makes the code much nicer, I like it!

**Pull request merged**